finalfrontier/src/skipgram_trainer.rs

```diff
@@ -211,13 +211,15 @@ where
             let right = cmp::min(self.i + context_size + 1, self.ids.len());
             let contexts = (left..right)
                 .filter(|&idx| idx != self.i)
-                .map(|idx| self.output_(self.ids[idx], self.i, idx))
+                .map(|idx| self.output_(self.ids[idx].single_idx() as usize, self.i, idx))
                 .fold(Vec::with_capacity(right - left), |mut contexts, idx| {
                     contexts.push(idx);
                     contexts
                 });
+            let mut word_idx = Idx::Single(self.ids[self.i].single_idx());
```

Comment:
This is hard to follow... Can't the iterator just continue to use `usize` for the focus token? Maybe just map the `Vec` to one of `usize` in the constructor?

Reply:
Do you mean `(usize, Vec<usize>` as `Item` for the iterator?

I'm not sure what you mean by mapping `Vec` and `usize`, get the single index from the indices `Vec` in the constructor and map those indices to the vectors?

Reply:
I went for the `mem::swap` and construction of a single idx in both proposed PRs because we can't move out of `self.ids` and we don't get `Copy` on the `Idx` types. So there would be a few options I guess:

* Store two id `Vec`s in the struct, one for the focus indices and one for the context indices.
* Store a single id `Vec`, don't touch the entries and `.clone()` the indices to return them.
* The proposed solution, construct a cheap alternative and swap it with the focus indices we need to return.

I think if we don't want to abstract over the index types in this struct, then we'd need to add an explicit indirection in `Trainer`, which was a solution I proposed quite some time ago with some `AssociatedIndices` trait that returned an empty slice for `SimpleVocab` and a slice with subword indices for `SubwordVocab`.

Reply:
I don't mean the iterator item, but the member `ids` member, which used to be `Vec<usize>` and is now `Vec<Idx>`. I thought the constructor could just `map` + `collect` the `Vec<Idx>` to `Vec<usize>` and use that. But I see now that the iterator-using code now uses `Multi`.

Reply:
I think that I now understand better why the code is difficult to read in places. `Idx` is very generic and `single_idx` does not really convey what is meant.

I think you'd rather want some name like `WordIndex` and methods could be `word_idx` and `subword_indices` (if it is the variant with subwords). But I will read the trait-based PR, because this really sounds like it should just be an associated type of the vocab.

In this particular case, a code comment describing what happens helps.